engine/src/main/java/org/pentaho/di/trans/steps/javafilter/JavaFilterMeta.java

```diff
@@ -0,0 +1,280 @@
+import org.pentaho.di.core.row.RowMetaInterface;
+import org.pentaho.di.core.util.Utils;
+import org.pentaho.di.core.variables.VariableSpace;
+import org.pentaho.di.core.xml.XMLHandler;
+import org.pentaho.di.i18n.BaseMessages;
+import org.pentaho.di.repository.ObjectId;
+import org.pentaho.di.repository.Repository;
+import org.pentaho.di.trans.Trans;
+import org.pentaho.di.trans.TransMeta;
+import org.pentaho.di.trans.step.BaseStepMeta;
+import org.pentaho.di.trans.step.StepDataInterface;
+import org.pentaho.di.trans.step.StepIOMeta;
+import org.pentaho.di.trans.step.StepIOMetaInterface;
+import org.pentaho.di.trans.step.StepInterface;
+import org.pentaho.di.trans.step.StepMeta;
+import org.pentaho.di.trans.step.StepMetaInterface;
+import org.pentaho.di.trans.step.errorhandling.Stream;
+import org.pentaho.di.trans.step.errorhandling.StreamIcon;
+import org.pentaho.di.trans.step.errorhandling.StreamInterface;
+import org.pentaho.di.trans.step.errorhandling.StreamInterface.StreamType;
+import org.pentaho.metastore.api.IMetaStore;
+import org.w3c.dom.Node;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Contains the meta-data for the java filter step: calculates conditions using Janino
+ *
+ * Created on 30-oct-2009
+ */
+public class JavaFilterMeta extends BaseStepMeta implements StepMetaInterface {
+  private static Class<?> PKG = JavaFilterMeta.class; // for i18n purposes, needed by Translator2!!
+
+  /** The formula calculations to be performed */
+  private String condition;
+
+  public JavaFilterMeta() {
+    super(); // allocate BaseStepMeta
+  }
+
+  public String getCondition() {
+    return condition;
+  }
+
+  public void setCondition( String condition ) {
+    this.condition = condition;
+  }
+
+  public void allocate( int nrCalcs ) {
+  }
+
+  public void loadXML( Node stepnode, List<DatabaseMeta> databases, IMetaStore metaStore ) throws KettleXMLException {
+    List<StreamInterface> targetStreams = getStepIOMeta().getTargetStreams();
+
+    targetStreams.get( 0 ).setSubject( XMLHandler.getTagValue( stepnode, "send_true_to" ) );
+    targetStreams.get( 1 ).setSubject( XMLHandler.getTagValue( stepnode, "send_false_to" ) );
+
+    condition = XMLHandler.getTagValue( stepnode, "condition" );
+  }
+
+  public String getXML() {
+    StringBuilder retval = new StringBuilder();
+
+    List<StreamInterface> targetStreams = getStepIOMeta().getTargetStreams();
+    retval.append( XMLHandler.addTagValue( "send_true_to", targetStreams.get( 0 ).getStepname() ) );
+    retval.append( XMLHandler.addTagValue( "send_false_to", targetStreams.get( 1 ).getStepname() ) );
+
+    retval.append( XMLHandler.addTagValue( "condition", condition ) );
+
+    return retval.toString();
+  }
+
+  public boolean equals( Object obj ) {
+    if ( obj != null && ( obj.getClass().equals( this.getClass() ) ) ) {
+      JavaFilterMeta m = (JavaFilterMeta) obj;
+      return ( getXML() == m.getXML() );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.